### PR TITLE
[SD-1319] As a Admin in Admin Panel -> Store Edit I can upload a Store Favicon

### PR DESCRIPTION
**Pull request by Vegann**

### CODE REVIEW

File: backend/app/controllers/spree/admin/stores_controller.rb [hunk 1]
```diff
@@ -58,7 +58,7 @@ def update
           redirect_to admin_stores_path
         else
           flash[:error] = "#{Spree.t('store_errors.unable_to_update')}: #{@store.errors.full_messages.join(', ')}"
-          render :edit
+          redirect_to edit_admin_store_path(@store)
```
Comment (damianlegawiec) [inline backend/app/controllers/spree/admin/stores_controller.rb:61]:
```suggestion
          redirect_to spree.edit_admin_store_path(@store)
```

It's always safer to use `spree` instance to scope the routes to Spree engine routes

File: backend/app/views/spree/admin/stores/_form.html.erb [hunk 2]
```diff
@@ -29,6 +29,16 @@
         <%= f.error_message_on :mailer_logo %>
       <% end %>
     <% end %>
+    <%= f.field_container :favicon_image, class: ['form-group'] do %>
+      <% if @store.favicon_image.attached? && @store.favicon_image.variable? %>
+        <%= image_tag main_app.url_for(@store.favicon_image.variant(resize: '104x104')) %>
+      <% end %>
+      <%= f.field_container :favicon_image, class: ['form-group'] do %>
+        <%= f.label :favicon_image, Spree.t(:favicon) %><br>
+        <%= f.file_field :favicon_image, accept: 'image/png, image/x-icon' %>
```
Comment (damianlegawiec) [inline backend/app/views/spree/admin/stores/_form.html.erb:38]:
```suggestion
        <%= f.file_field :favicon_image, accept: Spree::Store::FAVICON_ACCEPTED_FORMATS %>
```

File: core/app/models/spree/app_configuration.rb [hunk 3]
```diff
@@ -47,6 +47,7 @@ class AppConfiguration < Preferences::Configuration
     preference :disable_sku_validation, :boolean, default: false # when turned off disables the built-in SKU uniqueness validation
     preference :expedited_exchanges, :boolean, default: false # NOTE this requires payment profiles to be supported on your gateway of choice as well as a delayed job handler to be configured with activejob. kicks off an exchange shipment upon return authorization save. charge customer if they do not return items within timely manner.
     preference :expedited_exchanges_days_window, :integer, default: 14 # the amount of days the customer has to return their item after the expedited exchange is shipped in order to avoid being charged
+    preference :favicon, :string, default: 'favicon.ico'
```
Comment (damianlegawiec) [inline core/app/models/spree/app_configuration.rb:50]: @mad-eel let's drop the preference at all, it should be only manageable via Admin UI sorry I didn't tell you that before, we're moving away from app preferences to Store manageable configuration in the UI

File: core/app/models/spree/store.rb [hunk 4]
```diff
@@ -1,5 +1,8 @@
 module Spree
   class Store < Spree::Base
+    MAILER_LOGO_CONTENT_TYPES = ['image/png', 'image/jpg', 'image/jpeg'].freeze
+    FAVICON_CONTENT_TYPES = ['image/png', 'image/x-icon', 'image/vnd.microsoft.icon'].freeze
```
Comment (MatthewKennedy) [inline core/app/models/spree/store.rb:4]: You could probably get away with adding support for SVG to `FAVICON_CONTENT_TYPES`, although its not 100% supported by all browsers. https://caniuse.com/link-icon-svg

**Review by MatthewKennedy — APPROVED** [review]

**Review by damianlegawiec — COMMENTED** [review]
@mad-eel let's also expose the favicon path in the storefront serializer - https://github.com/spree/spree/blob/master/api/app/serializers/spree/v2/storefront/store_serializer.rb#L8

File: api/app/serializers/spree/v2/storefront/store_serializer.rb [hunk 5]
```diff
@@ -10,6 +10,13 @@ class StoreSerializer < BaseSerializer
 
         has_many :menus
         has_one :default_country, serializer: :country, record_type: :country, id_method_name: :default_country_id
+
+        attribute :favicon_path do |store|
+          if store.favicon_image.attached?
+            url_helpers = Rails.application.routes.url_helpers
+            url_helpers.rails_representation_path(store.favicon, only_path: true)
```
Comment (mad-eel) [inline api/app/serializers/spree/v2/storefront/store_serializer.rb:17]: @damianlegawiec I'm not sure about use cases, but do you think that relative path will be enough?

Comment (damianlegawiec) [reply api/app/serializers/spree/v2/storefront/store_serializer.rb:17]: @mad-eel it will be fine! :)

**Review by damianlegawiec — COMMENTED** [review]
@mad-eel can we also add a frontend spec testing the proper HTML with favicon generated when supplied?

**Review by damianlegawiec — APPROVED** [review]

**Pull request merged**